add bq/fnac device support

pazos opened the pull request:
Since those devices are almost copies of kobos (w/softfloat eabi, without simd) most code is almost there.

~~I refactored common kobo/bq code to input-ntx.h~~ Nah. bq devices don't wake up when power is applied via usb.

It builds ok w/ koxtoolchain cervantes target or with a ubuntu/linaro toolchain (given that gcc > 4.8 and glibc <= 2.13)

TODO:
- FBInk needs a bump. (current binary works, but breaks stock application on some circumstances)
~~build a Docker image, need help on that.~~ thanks @Frenzie

NiLuJe commented:
(next FBInk release probably sometime next week, because I'm expecting doom & misery with the Forma and its magic rotation handling...).

Frenzie commented:
@pazos On Ubuntu you can just install Docker and for the rest you can find the relevant commands in the Makefiles and scripts.

https://github.com/koreader/virdevenv/blob/13c37dcc21d6e8acffe53e0d88e2a6e3562bd169/docker/ubuntu/kokobo/Makefile

You can probably just copy the Kobo almost wholesale.

Uploading it to Docker cloud is a touch trickier in the sense that you need an account, but otherwise also covered by the Makefile. :-)

pazos commented:
Pinging @Frenzie @NiLuJe for review.

NiLuJe commented:
What happens, exactly, with FBInk and the stock app?

NiLuJe commented on input/input-cervantes.h:
```diff
@@ -0,0 +1,66 @@
+
+    re = ue_init_listener(&listener);
+    if (re < 0) {
+        fprintf(stderr, "[ko-input]: Failed to initilize libue listener, err: %d\n", re);
```
initialize ;)

NOTE: It's not actually your typo, so if you can fix the same one in input-kobo while you're there, that'd be great ;).

pazos replied:
do you mean re > 0 ?

NiLuJe replied:
No, I meant `initilize` -> `initialize` in the fprintf ;).

pazos replied:
ah, that makes sense :+1:

NiLuJe commented:
Generally LGTM otherwise, besides the small inherited typo ;).

NiLuJe commented:
After a quick check, nope, the 0x0 region check looks solid, and Cervantes builds shouldn't bypass it.

The only other thing I can think of is that something doesn't like clear's memset on smem_len, so you can try to replace https://github.com/NiLuJe/FBInk/blob/8d6c857eccf2dec0aa974a8837d355c6e14fa9f5/fbink.c#L458 by ```memset(fbPtr, v, fInfo.line_length * vInfo.yres_virtual);```, and see if that helps?

NiLuJe commented:
Did that anyway in https://github.com/NiLuJe/FBInk/commit/0ae5dbcc13278178aee4178499e0eb2f9776003b, because it can't hurt ;).

pazos commented:
@NiLuJe: it seems that fbink fails on travis because we need to specify C99. That happens when using linaro toolchains but not using koxtoolchain. Pinging @Frenzie too because it seems that the docker image is relying on a toolchain other than arm-cervantes-linux-gnueabi.

pazos commented:
> Did that anyway in NiLuJe/FBInk@0ae5dbc, because it can't hurt ;).

Will test that later. I debootstrapped a new rootfs just for the fun and have no access to stock reader app now :see_no_evil:

Frenzie commented:
> Pinging @Frenzie too because it seems that the docker image is relying on a toolchain other than arm-cervantes-linux-gnueabi.

I merely built and uploaded the Docker image you provided. ;-)

I can sh into the Docker image this weekend, but for the moment I'm more suspicious of your Makefile here.

pazos commented:
> I can sh into the Docker image this weekend, but for the moment I'm more suspicious of your Makefile here.

Not sure if it works but I [opened a PR in virdevenv](https://github.com/koreader/virdevenv/pull/17).

pazos commented:
I reverted the mxcfb headers to match the behaviour of stock reader app (which is pretty conservative and does the same for all devices, see:https://github.com/bq/cervantes-qt/blob/eink-imx508/src/plugins/gfxdrivers/einkfb/linux/mxcfb.h) and squashed commits.

NiLuJe commented:
Okay, after taking a look at https://github.com/bq/cervantes-qt/blob/eink-imx508/src/plugins/gfxdrivers/einkfb/einkfb.cpp, I can see how my memset would potentially break Qt's little mind ;).

The current approach probably still does, I'll see about that tomorrow...

NiLuJe commented:
The clear screen issue *should* be taken care of in https://github.com/NiLuJe/FBInk/commit/10beff8d99d3ef359274ebb5ad502e4123f3554d ;).

Frenzie commented:
@pazos @NiLuJe Is this PR done now?

pazos commented:
@Frenzie: I would like to submit a couple more changes in the waveform_modes table since current behaviour with new devices is far from perfect. Give me a couple of hours/days :smile:

pazos commented:
pinging @NiLuJe about the REAGL stuff. I tried a bunch of flags based on what we are doing on other devices but the result is not good. Also tried newer mxc_wait_for_update_complete ioctl calls without success (I mean it works but doesn't improve anything, making the UI feels slower).

@Frenzie from my end I'm done with this PR, so if you think it is ready to be merged then I will focus on submit a proper frontend stuff.

NiLuJe commented:
Yeah, I doubt you could get anything great with explicit REAGL calls, only the Aura did on Kobo's side, and even then, it was fairly weird.

Frenzie commented on Makefile.defs:
```diff
@@ -286,6 +297,13 @@ else ifeq ($(TARGET), android)
+else ifeq ($(TARGET), cervantes)
+	ARM_ARCH:=$(ARMV7_GENERIC_ARCH)
+	ARM_ARCH+=-mfloat-abi=softfp
+    ifeq ($(shell PATH='$(PATH)' $(CC) -dumpmachine 2>/dev/null), arm-linux-gnueabi)
```
inconsistent indentation

You generally shouldn't use spaces in Makefiles.

pazos replied:
same identation as other devices, like kindlepw2, What I should do?

Frenzie replied:
It's not. The other devices use tab as they should while these here are spaces.

pazos replied:
@Frenzie Could you please point me to your text editor?. I opened in vim. First two lines are tabs, the ifeq/endif stuff just spaces. The prstux stuff contains spaces too.

Can't fix an error if I can't see it :cry:

NiLuJe replied:
The kindle & kindlepw2 indents are actually also wrong, which might be the source of the confusion ;).

Note that the space/tab distinction is actually a bit of a mess in here.
I wholeheartedly agree with @Frenzie, which is why i think the tabs are mine, and the spaces are whatever was there before ;^.

NiLuJe replied:
See the android case for one done right ;).

Frenzie replied:
@pazos 
> First two lines are tabs, the ifeq/endif stuff just spaces. The prstux stuff contains spaces too.

So prstux is wrong too. ;-) But please leave fixing the others for a separate PR.

@NiLuJe 
> See the android case for one done right ;).

By pure coincidence that's the one I looked at. (It also happens to be the one I'm responsible for.)

pazos replied:
No problem, reverted Makefile.defs and squashed again. :+1:

Frenzie commented on thirdparty/glib/CMakeLists.txt:
```diff
@@ -23,7 +23,7 @@ ep_get_binary_dir(BINARY_DIR)
 # sometimes just hoping there's a python binary somewhere in $PATH (which, granted, there should be)...
 # Enforce that to python3 for containers where that's the only available Python version.
-if(($ENV{KINDLE}) OR ($ENV{KOBO}) OR ($ENV{APPIMAGE}))
+if(($ENV{KINDLE}) OR ($ENV{KOBO}) OR ($ENV{CERVANTES}) OR ($ENV{APPIMAGE}))
```
Ah, I'm too blame for putting AppImage at the end. Could you please fix it?

Frenzie commented on thirdparty/libjpeg-turbo/CMakeLists.txt:
```diff
@@ -31,9 +31,9 @@ set(CFG_OPTS "${CFG_OPTS} -DCMAKE_C_COMPILER='${CMAKE_C_COMPILER}' -DCMAKE_C_COM
 
 # c.f., http://trac.ak-team.com/trac/browser/niluje/Configs/trunk/Kindle/Misc/CMakeCross.txt
-if((DEFINED ENV{KINDLE}) OR (DEFINED ENV{LEGACY}) OR (DEFINED ENV{KOBO}) OR (DEFINED ENV{POCKETBOOK}) OR (DEFINED ENV{UBUNTUTOUCH}) OR (DEFINED ENV{SONY_PRSTUX}))
+if((DEFINED ENV{KINDLE}) OR (DEFINED ENV{LEGACY}) OR (DEFINED ENV{KOBO}) OR (DEFINED ENV{POCKETBOOK}) OR (DEFINED ENV{UBUNTUTOUCH}) OR (DEFINED ENV{SONY_PRSTUX}) OR (DEFINED ENV{CERVANTES}))
```
Just adding things to the end works up to the point where it becomes illegible. ;-)

pazos replied:
switching to two lines then

Frenzie replied:
I mean alphabetical; my text editor auto-wraps just fine. :-)

Frenzie commented on thirdparty/libjpeg-turbo/CMakeLists.txt:
```diff
@@ -31,9 +31,9 @@ set(CFG_OPTS "${CFG_OPTS} -DCMAKE_C_COMPILER='${CMAKE_C_COMPILER}' -DCMAKE_C_COM
+if((DEFINED ENV{KINDLE}) OR (DEFINED ENV{LEGACY}) OR (DEFINED ENV{KOBO}) OR (DEFINED ENV{POCKETBOOK}) OR (DEFINED ENV{UBUNTUTOUCH}) OR (DEFINED ENV{SONY_PRSTUX}) OR (DEFINED ENV{CERVANTES}))
 	set(CFG_OPTS "${CFG_OPTS} -DCMAKE_SYSTEM_NAME='Linux'")
-endif((DEFINED ENV{KINDLE}) OR (DEFINED ENV{LEGACY}) OR (DEFINED ENV{KOBO}) OR (DEFINED ENV{POCKETBOOK}) OR (DEFINED ENV{UBUNTUTOUCH}) OR (DEFINED ENV{SONY_PRSTUX}))
+endif((DEFINED ENV{KINDLE}) OR (DEFINED ENV{LEGACY}) OR (DEFINED ENV{KOBO}) OR (DEFINED ENV{POCKETBOOK}) OR (DEFINED ENV{UBUNTUTOUCH}) OR (DEFINED ENV{SONY_PRSTUX}) OR (DEFINED ENV{CERVANTES}))
```
Here too.

Frenzie commented:
Thanks!
